### ISSUE TITLE
[Redshift] Update certain columns values

### ISSUE THREAD

**Issue by Olaktal** [embulk/embulk-output-jdbc#139]
Hello,

I need to update the value of multiple columns in a table and I'm not sure I can do it for now so maybe I'm asking for a possible evolution. Here is an example of what I need:

```
create table questions ( question_id int IDENTITY(1,1), title varchar(100),content varchar(1000),average_answer_time int)

insert into questions (title,content) VALUES ('title1', 'content1'),('title2', 'content2'),('title3', 'content3')
```

I have to add data to the table rows but I can't do it during the first import for multiple factors (not available at that time, in another database or format, ...). So I need to update multiple rows afterwards.
Here is a simple example of input data:

![image](https://cloud.githubusercontent.com/assets/5711131/19483276/1992ffc8-9554-11e6-8fe4-f99c7a8b1cc8.png)

Basically, the out part of the YML file would be like that:

```
out:
  type: redshift
  host: ***
  user: ***
  password: ***
  database: ***
  schema: ***
  table: questions
  access_key_id: ***
  secret_access_key: ***
  iam_user_name: ***
  s3_bucket: ***
  s3_key_prefix: ***
  mode: update
  update_keys: ["question_id "] (Could have multiple update_keys)
  column_updated: 
    average_answer_time : {value_type: int}
   where:  (not used here but it can be usefull, I already have use cases in my head)
```

Is it possible @hito4t  ?
I suppose this plugin will be considered as the most evolved plugin in embulk with this functionality ahah

Best regards,
Vincent

**Comment by hito4t** [comment]
Hi,
embulk-output-xxx has `merge` mode and `merge_direct` mode to update or insert rows.
But, unfortunately, the following yml doesn't work now.
<pre>
in:
  type: file
  path_prefix: 'questions.csv'
  parser:
    charset: UTF-8
    newline: CRLF
    type: csv
    delimiter: ','
    header_line: false
    columns:
    - {name: question_id, type: long}
    - {name: average_answer_time, type: long}
out:
    type: redshift
    ...
    table: questions
    mode: merge
    merge_keys: [question_id]
</pre>

embulk-output-redshift `merge` mode works as follows.
1. insert rows into temporary table
2. delete existing rows from target table
3. insert rows into target table from temporary table

But 3. doesn't work because `questions` table has identity column.

`merge_mode` directly inserts or updates rows in target table, but embulk-output-redshift doesn't support the mode because Redshift doesn't have SQL which can insert or update a row in one statement.
In general, we can do that by multiple JDBC calls. But embulk-output-jdbc doesn't support that because performance may decrease significantly. 
But it would be better to work slowly than not to work.

Do you have any ideas?

**Comment by Olaktal** [comment]
Thanks for your answer @hito4t !

I understand the issue and it occurs when you want to insert but my request was only about updating data, not inserting :)

Look at the following example -->

Data after insert in my table is:
![image](https://cloud.githubusercontent.com/assets/5711131/19566591/248de724-96eb-11e6-8816-28b47cb122d5.png)

My input data is:
![image](https://cloud.githubusercontent.com/assets/5711131/19566598/2e530faa-96eb-11e6-9113-7f8cc76928a2.png)
(as described above)

And the result should be;
![image](https://cloud.githubusercontent.com/assets/5711131/19566632/5211e0ec-96eb-11e6-942e-5190f3a46f8e.png)

No need to insert dat I just want to update, column 'question_id' can be identity or not, It doesn't matter I guess. An SQL update statement should be enougth to do that that's why I called the mode 'update'.

Unfortunately, I didn't noticed the existing rows where deleted and it is a BIG issue for me because when I insert data, I linked them to another table (a FACT table) and I don't want my IDs to change !

I'll answer to you back tomorrow morning about this issue. (France time)

Thanks
Vincent

**Comment by Olaktal** [comment]
Yes I guess we need to execute the 3 SQL statement as follow :

UPDATE _table_name_ 
JOIN _tmp_table1_
SET _table_name_ .col1 = _tmp_table1_.col1, ... (all columns except merge keys and Identity column)
WHERE _table_name_._merge_key_ = _tmp_table1_._merge_key_ (could be multiple merge keys)

UPDATE _table_name_ 
JOIN _tmp_table2_
SET _table_name_ .col1 = _tmp_table2_.col1, ... (all columns except merge keys and Identity column)
WHERE _table_name_._merge_key_ = _tmp_table2_._merge_key_ (could be multiple merge keys)

INSERT INTO dim_client (_column_list_except_identity_column_)
(SELECT _column_list_ FROM _tmp_table1_ WHERE NOT EXISTS (SELECT 1 FROM _table_name_ WHERE _table_name_._merge_key_ = _tmp_table1_._merge_key_)
UNION
(SELECT _column_list_ FROM _tmp_table2_ WHERE NOT EXISTS (SELECT 1 FROM _table_name_ WHERE _table_name_._merge_key_ = _tmp_table2_._merge_key_)
)

Do you think it is possible to make it work like this ?
As you said, performance may be poor but it is better done than not ...

Tell me if you need anything.

Thanks
Vincent

**Comment by Olaktal** [comment]
Hi @hito4t,

I hope you had a nice weekend :)

I don't know how to commit my code but the following code looks like it could solve the problem. (can contains errors, didn't test it)

```
   @Override
    protected String buildCollectMergeSql(List<String> fromTables, JdbcSchema schema, String toTable, MergeConfig mergeConfig) throws SQLException
    {
        StringBuilder sb = new StringBuilder();

        sb.append("BEGIN TRANSACTION;");

        sb.append("UPDATE ");
        quoteIdentifierString(sb, toTable);
        sb.append(" SET ");
        for (int i = 0; i < schema.getCount(); i++) {
            if (i != 0) { sb.append(", "); }
            quoteIdentifierString(sb, schema.getColumnName(i));
            sb.append(" = ");
            sb.append("S.");
            quoteIdentifierString(sb, schema.getColumnName(i));
        }
        sb.append(" FROM ( ");
        for (int i = 0; i < fromTables.size(); i++) {
            if (i != 0) { sb.append(" UNION ALL "); }
            sb.append(" SELECT ");
            for (int j = 0; j < schema.getCount(); j++) {
                if (j != 0) { sb.append(", "); }
                quoteIdentifierString(sb, schema.getColumnName(j));
            }
            sb.append(" FROM ");
            quoteIdentifierString(sb, fromTables.get(i));
        }
        sb.append(" ) S WHERE ");

        List<String> mergeKeys = mergeConfig.getMergeKeys();
        for (int i = 0; i < mergeKeys.size(); i++) {
            if (i != 0) { sb.append(" AND "); }
            sb.append("S.");
            quoteIdentifierString(sb, mergeKeys.get(i));
            sb.append(" = ");
            quoteIdentifierString(sb, toTable);
            sb.append(".");
            quoteIdentifierString(sb, mergeKeys.get(i));
        }
        sb.append(";");

        sb.append("INSERT INTO ");
        quoteIdentifierString(sb, toTable);
        sb.append(" (");
        for (int i = 0; i < schema.getCount(); i++) {
            if (i != 0) { sb.append(", "); }
            quoteIdentifierString(sb, schema.getColumnName(i));
        }
        sb.append(") (");
        for (int i = 0; i < fromTables.size(); i++) {
            if (i != 0) { sb.append(" UNION ALL "); }
            sb.append("(SELECT ");
            for (int j = 0; j < schema.getCount(); j++) {
                if (j != 0) { sb.append(", "); }
                quoteIdentifierString(sb, schema.getColumnName(j));
            }
            sb.append(" FROM ");
            quoteIdentifierString(sb, fromTables.get(i));
            sb.append(" WHERE NOT EXISTS (SELECT 1 FROM ");
            quoteIdentifierString(sb, fromTables.get(i));
            sb.append(" WHERE ");
            List<String> mergeKeys = mergeConfig.getMergeKeys();
            for (int i = 0; i < mergeKeys.size(); i++) {
                if (i != 0) { sb.append(" AND "); }
                quoteIdentifierString(sb, fromTables.get(i));
                sb.append(".");
                quoteIdentifierString(sb, mergeKeys.get(i));
                sb.append(" = ");
                quoteIdentifierString(sb, toTable);
                sb.append(".");
                quoteIdentifierString(sb, mergeKeys.get(i));
            }
            sb.append(") ");
        }
        sb.append(");");

        sb.append("END TRANSACTION;");

        return sb.toString();

}
```

**Comment by hito4t** [comment]
Hi,
Thank you for your idea!

I've tried the code after correcting as follows.
<pre>
    @Override
    protected String buildCollectMergeSql(List<String> fromTables, JdbcSchema schema, String toTable, MergeConfig mergeConfig) throws SQLException
    {
        List<String> mergeKeys = mergeConfig.getMergeKeys();
        StringBuilder sb = new StringBuilder();

        sb.append("BEGIN TRANSACTION;");

        sb.append("UPDATE ");
        quoteIdentifierString(sb, toTable);
        sb.append(" SET ");
        List<String> updateKeys = new ArrayList<String>();
        for (int i = 0; i < schema.getCount(); i++) {
            String updateKey = schema.getColumnName(i);
            if (!mergeKeys.contains(updateKey)) {
                updateKeys.add(updateKey);
            }
        }

        for (int i = 0; i < updateKeys.size(); i++) {

        ...

           sb.append(" WHERE ");
            for (int j = 0; j < mergeKeys.size(); j++) {
                if (j != 0) { sb.append(" AND "); }
                quoteIdentifierString(sb, fromTables.get(i));
                sb.append(".");
                quoteIdentifierString(sb, mergeKeys.get(j));
                sb.append(" = ");
                quoteIdentifierString(sb, toTable);
                sb.append(".");
                quoteIdentifierString(sb, mergeKeys.get(j));
            }
            sb.append(")) ");
        ...
</pre>

But, unfortunately, it doesn't work.
INSERT with IDENTITY columns causes error even if no record was actually inserted.

IDENTITY columns are troublesome.
I'm looking for another way...

Thanks.

**Comment by Olaktal** [comment]
Hi @hito4t,

Oh I didn't know mergeKeys was inside schema column, you did great with your updateKeys list :)
Can you please give me the steps to test my code once I have updated the RedshiftOutputConnection.java file in the folder ~/.embulk/jruby/2.3.0/gems/embulk-output-redshift-0.6.5/src/main/java/org/embulk/output/redshift/ ? 
Maybe I will be more helpfull if I can test my code myself :)

I don't have your full code function but I don't understand why it doesn't works.
Here was my SQL example:
```
create table tmp1 (original_id int, txt varchar(100));
create table tmp2 (original_id int,  txt varchar(100));
create table tmp3 (id int identity(1,1) ,original_id int,  txt varchar(100));

insert into tmp1 VALUES (1,'plop1');
insert into tmp2 VALUES (2,'plop2');
insert into tmp2 VALUES (3,'plop3');

insert into tmp3(original_id,txt)  VALUES (1,'p1');
insert into tmp3(original_id,txt)  VALUES (2,'p2');

UPDATE tmp3
SET txt = tmp4.txt
FROM (SELECT original_id,txt FROM tmp1 UNION SELECT original_id,txt FROM tmp2) tmp4
WHERE tmp3.original_id = tmp4.original_id

INSERT INTO tmp3(original_id,txt)
(SELECT original_id, txt FROM tmp1 WHERE NOT EXISTS (SELECT 1 FROM tmp3 WHERE tmp3.original_id = tmp1.original_id)
UNION 
SELECT original_id, txt FROM tmp2 WHERE NOT EXISTS (SELECT 1 FROM tmp3 WHERE tmp3.original_id = tmp2.original_id))
```
Update and Insert work perfectly well  in this example.

Good luck and keep me posted :)

Vincent

**Comment by hiroyuki-sato** [comment]
Hello @Olaktal 
It it means that you want to know how to build embulk-ouptut-redshift?

If so, Please try the following step.

```
git clone https://github.com/embulk/embulk-output-jdbc
cd embulk-output-jdbc
git checkout -b redshift_merge_columns origin/redshift_merge_columns
```

Modify code.

compile 

```
./gradlew classpath
```

test

```
embulk run -l debug -I /path/to/embulk-output-jdbc/embulk-output-redshift/lib/ conf.yml
```

**Comment by Olaktal** [comment]
Hello @hiroyuki-sato,

Thanks It worked ! But I got the following error with git checkout 

```
fatal: Cannot update paths and switch to branch 'redshift_merge_columns' at the same time.
Did you intend to checkout 'origin/redshift_merge_columns' which can not be resolved as commit?
```

@hito4t The following code is tested (thx to @hiroyuki-sato) and works perfectly for me ! Please test it and tell me if you can release it :)

```
    @Override
    protected String buildCollectMergeSql(List<String> fromTables, JdbcSchema schema, String toTable, MergeConfig mergeConfig) throws SQLException
    {
        StringBuilder sb = new StringBuilder();

        List<String> mergeKeys = mergeConfig.getMergeKeys();

        List updateKeys = new ArrayList();
        for (int i = 0; i < schema.getCount(); i++) {
            String updateKey = schema.getColumnName(i);
            if (!mergeKeys.contains(updateKey)) {
                updateKeys.add(updateKey);
            }
        }

        sb.append("BEGIN TRANSACTION;");

        sb.append("UPDATE ");
        quoteIdentifierString(sb, toTable);
        sb.append(" SET ");
        for (int i = 0; i < updateKeys.size(); i++) {
            if (i != 0) { sb.append(", "); }
            sb.append(updateKeys.get(i));
            sb.append(" = ");
            sb.append("S.");
            sb.append(updateKeys.get(i));
        }
        sb.append(" FROM ( ");
        for (int i = 0; i < fromTables.size(); i++) {
            if (i != 0) { sb.append(" UNION ALL "); }
            sb.append(" SELECT ");
            for (int j = 0; j < schema.getCount(); j++) {
                if (j != 0) { sb.append(", "); }
                quoteIdentifierString(sb, schema.getColumnName(j));
            }
            sb.append(" FROM ");
            quoteIdentifierString(sb, fromTables.get(i));
        }
        sb.append(" ) S WHERE ");

        for (int i = 0; i < mergeKeys.size(); i++) {
            if (i != 0) { sb.append(" AND "); }
            sb.append("S.");
            quoteIdentifierString(sb, mergeKeys.get(i));
            sb.append(" = ");
            quoteIdentifierString(sb, toTable);
            sb.append(".");
            quoteIdentifierString(sb, mergeKeys.get(i));
        }
        sb.append(";");

        sb.append("INSERT INTO ");
        quoteIdentifierString(sb, toTable);
        sb.append(" (");
        for (int i = 0; i < schema.getCount(); i++) {
            if (i != 0) { sb.append(", "); }
            quoteIdentifierString(sb, schema.getColumnName(i));
        }
        sb.append(") (");
        for (int i = 0; i < fromTables.size(); i++) {
            if (i != 0) { sb.append(" UNION ALL ("); }
            sb.append("SELECT ");
            for (int j = 0; j < schema.getCount(); j++) {
                if (j != 0) { sb.append(", "); }
                quoteIdentifierString(sb, schema.getColumnName(j));
            }
            sb.append(" FROM ");
            quoteIdentifierString(sb, fromTables.get(i));
            sb.append(" WHERE NOT EXISTS (SELECT 1 FROM ");
            quoteIdentifierString(sb, toTable);
            sb.append(" WHERE ");

            for (int k = 0; k < mergeKeys.size(); k++) {
                if (k != 0) { sb.append(" AND "); }
                quoteIdentifierString(sb, fromTables.get(i));
                sb.append(".");
                quoteIdentifierString(sb, mergeKeys.get(k));
                sb.append(" = ");
                quoteIdentifierString(sb, toTable);
                sb.append(".");
                quoteIdentifierString(sb, mergeKeys.get(k));
            }
            sb.append(")) ");
        }
        sb.append(";");

        sb.append("END TRANSACTION;");

        return sb.toString();

    }

}
```

@hiroyuki-sato Is it possible for someone like me to release code ? If yes, could you please give me the steps ?

**Comment by hiroyuki-sato** [comment]
Hello @Olaktal 
Oh sorry, It's the typo.

You don't need the following part.
~~git checkout -b redshift_merge_columns origin/redshift_merge_columns~~

Let's fork embulk-output-jdbc. (Just press fork button.)
You will find `Fork` button on the right of the top.
https://github.com/embulk/embulk-output-jdbc

Then, you will find in your repository.
https://github.com/Olaktal/embulk-output-jdbc

clone it.

```
git clone https://github.com/Olaktal/embulk-output-jdbc
```

create the new branch 

```
git checkout -b redshift_merge
```

modify it

commit it.

```
git commit .
```

```
git push origin redshift_merge
```

Then we can test it.

**Comment by hito4t** [comment]
Hi @Olaktal 
I tested with old DDL, then it didn't work.
<pre>
create table questions ( 
  question_id int IDENTITY(1,1), 
  title varchar(100),
  content varchar(1000),
  average_answer_time int)
</pre>

Now I've tested with new DDL, then it worked!
<pre>
create table questions ( 
  question_id int IDENTITY(1,1), 
  original_id int, 
  title varchar(100),
  content varchar(1000),
  average_answer_time int)
</pre>

Input csv:
<pre>
original_id,title
1,xxx
2,yyy
9,zzz
</pre>

yml:
<pre>
in:
  type: file
  path_prefix: 'data/questions.csv'
  parser:
    charset: UTF-8
    newline: CRLF
    type: csv
    delimiter: ','
    header_line: true
    columns:
    - {name: original_id, type: long}
    - {name: title, type: string}
out:
    type: redshift
    host: xxx
    port: 5439
    database: embulk
    access_key_id: xxx
    secret_access_key: xxx
    iam_user_name: xxx
    s3_bucket: xxx
    s3_key_prefix: xxx
    user: xxx
    password: xxx
    table: questions
    mode: merge
    merge_keys: [original_id]
</pre>

Current `RedshiftOutputConnection` generates DELETE statement, so columns not existed in input schema will be cleared.

Your `RedshiftOutputConnection` generates UPDATE statement, so columns not existed in input schema won't be changed. It is very nice!

I'll execute some more test cases (including performance test), and I'll merge if no problems.

Thanks!

**Comment by Olaktal** [comment]
Hi @hiroyuki-sato  and @hito4t ,

@hito4t --> Yes Merge key cannot be the IDENTITY column :)
 I thought about a lot of SQL statement before getting this one worked properly and efficiently.

I push my code, as @hiroyuki-sato described , so now I know all the steps to contribute :)

Merge function patched, I'll take a look at my first question about the "update only" mode but the modifications sound a bit harder because it seems I need to modify some files in the embulk-output-jdbc plugin ... (add a new mode,  new config ...)

Keep me posted ;)

Thanks
Vincent

**Comment by Olaktal** [comment]
Hello @hito4t ,

I'm trying to dev the "update_only" mode but I got the following error on my code when gradlewing

```
ubuntu@ip-172-31-42-201:~/embulk-output-jdbc$ ./gradlew classpath
:embulk-output-jdbc:compileJava
/home/ubuntu/embulk-output-jdbc/embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java:212: error: cannot find symbol
        @JsonPropertysetIgnoreMergeKeys
         ^
  symbol:   class JsonPropertysetIgnoreMergeKeys
  location: class Features
/home/ubuntu/embulk-output-jdbc/embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java:225: error: cannot find symbol
        @JsonPropertysetIgnoreUpdateKeys
         ^
  symbol:   class JsonPropertysetIgnoreUpdateKeys
  location: class Features
/home/ubuntu/embulk-output-jdbc/embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java:622: error: cannot find symbol
                task.setUpdateKeys(Optional.<List<String>>of(ImmutableList.<String>of()));
                    ^
  symbol:   method setUpdateKeys(Optional<List<String>>)
  location: variable task of type PluginTask
/home/ubuntu/embulk-output-jdbc/embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java:639: error: cannot find symbol
                task.setUpdateKeys(Optional.<List<String>>of(builder.build()));
                    ^
  symbol:   method setUpdateKeys(Optional<List<String>>)
  location: variable task of type PluginTask
/home/ubuntu/embulk-output-jdbc/embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java:646: error: cannot find symbol
            task.setUpdateKeys(Optional.<List<String>>absent());
                ^
  symbol:   method setUpdateKeys(Optional<List<String>>)
  location: variable task of type PluginTask
/home/ubuntu/embulk-output-jdbc/embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java:785: error: cannot find symbol
            con.collectUpdate(task.getIntermediateTables().get(), schema, task.getActualTable(), new UpdateConfig(task.getUpdateKeys().get(), task.getUpdateRule()));
                                                                                                                                                  ^
  symbol:   method getUpdateRule()
  location: variable task of type PluginTask
/home/ubuntu/embulk-output-jdbc/embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java:935: error: cannot find symbol
            } else if (task.getMode() == Mode.UPDATE_DIRECT) {
                                             ^
  symbol:   variable UPDATE_DIRECT
  location: class Mode
/home/ubuntu/embulk-output-jdbc/embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java:936: error: cannot find symbol
                configUpdate = Optional.of(new UpdateConfig(task.getUpdateKeys().get(), task.getUpdateRule()));
                                                                                            ^
  symbol:   method getUpdateRule()
  location: variable task of type PluginTask
8 errors
:embulk-output-jdbc:compileJava FAILED

FAILURE: Build failed with an exception.
```

Can you indicate me where does the problem come from ? Maybe I forgot to modify a file ?
I created a branch called "redshift_update_mode" with all my files.
https://github.com/Olaktal/embulk-output-jdbc

Remember, I want this update mode because I only want an update in my SQL statement. I can't take the risk a new rows is inserted by using merge mode in case my input contains a new row.

Can you help me on it ?

Thanks
Vincent

**Comment by hito4t** [comment]
Hi @Olaktal ,

I understand you should only update existing rows, but we should deliberate before adding a new mode because this plugin is used by many users.

Generally speaking, the plugin shouldn't ignore input records.
I think when an input record doesn't exist in the target table, the plugin should
- throw an exception and abort.

or

- write the record to an another table or a file specified by yml.

or

- write the record to the log file (but it may become too big).

Another idea is that the plugin executes SQL specified by yml after loading (as embulk-input-jdbc has `after_select` feature).
By the feature, you can load records to a work table, then update the target table by the work table.


Anyway, the compile errors are not caused by your environment, but by your code itself.

> @JsonPropertysetIgnoreMergeKeys

@JsonProperty or @JsonProperty("setIgnoreMergeKeys")?


> task.setUpdateKeys(Optional.<List<String>>of(ImmutableList.<String>of()));

`PluginTask` doesn't have `setUpdateKeys` method.

>             con.collectUpdate(task.getIntermediateTables().get(), schema, task.getActualTable(), new UpdateConfig(task.getUpdateKeys().get(), task.getUpdateRule()));

`PluginTask` doesn't have `getUpdateRule` method.


>             } else if (task.getMode() == Mode.UPDATE_DIRECT) {

`Mode` has `UPDATE`, not `UPDATE_DIRECT` .

Thanks!

**Comment by Olaktal** [comment]
Hello @hito4t,

I totally agree with you. We can still leave the choice to the user: If he specify a 'log_table' it will store inside, else if he specify the name of the file it will store ... else abort.

Maybe the best choice would be to create a 'log_table' anyway because:
- It can scale.
- You can query it and make analysis easier.
- You won't have a lot of (big) files.
- All YML file can store into a single table. (we don't want XX log table to be created during a night batch, It will be a real mess)

The log_table design could be like:
```
CREATE TABLE log_table (target_table varchar(100), create_time timestamp, row text)
/* example of row value: update_key1_col_name = value | update_col1_name = value | update_col2_name = value | ... */
```

But I'm not sure I have the skill/knowledge to do that, I barely know how to add a new mode :D
For me the current process is:
input --> S3 (--> TMP tables (if not a direct mode)) -->Target table

We can't do update with copy function so we have to store the input into intermediate tables.
My guess is the best way to do it is:

input --> S3 --> TMP tables -> Update the target table --> Create/insert into 'log_table' if input records are not in target table

The SQL statement looks like a lot of what I did for the merge mode.

About the after_select idea, to me it is a SQL statement you can execute in the input database but I have a read only user in the input database so I don't think I can do anything ...

What do you think ?

Vincent
Thanks

**Comment by hito4t** [comment]
Hi @Olaktal ,

I can't comment about `update` mode right away because it is a not easy issue.
I'm sorry but I'll comment later.

> About the after_select idea, to me it is a SQL statement you can execute in the input database but I have a read only user in the input database so I don't think I can do anything ...

My idea is adding the new feature `after_load` to embulk_output_jdbc.
embulk_output_jdbc loads records, then executes SQL specified in yml using the connection defined for loading.
The SQL will be able to update the target table. 

Thanks

**Comment by Olaktal** [comment]
Oh ok I understand, great idea !

I see only two limitations:
- it is less efficient than the update mode because you have to insert data into a table first
- you won't have the log (unless you design and load yourself a table during the after_load statement)

But it is easier to set up into embulk_output_jdbc I think :)

**Comment by hito4t** [comment]
Hi @Olaktal ,
I've started to implementing the `after_load` feature.

> it is less efficient than the update mode because you have to insert data into a table first

The update mode will also load records to a temporary table first, then update the target table, won't it?

**Comment by Olaktal** [comment]
Hi @hito4t,

Maybe it depends on how you will design after_load but for me the path of data with after_load will be like:
``` Input (--> s3) --> TMP tables --> To_table -->update statement (UPDATE table_name ...)```

the update will be instead:
```Input (--> s3) --> TMP tables -->update statement (UPDATE table_name ...)```

Because we can access the TMP tables directly in the code and not in the YML.

**Comment by hito4t** [comment]
Hi @Olaktal ,
You can use `after_load` with any mode.
If you use `insert_direct` mode, only one intermediate table will be used, because `insert_direct` mode doesn't generate TMP tables.
`Input (--> s3) --insert_direct--> work_table  --> after_load (UPDATE target_table ... SELECT ... FROM work_table ...)`

yml will be as follows.
<pre>
out:
    type: redshift
    ...
    table: work_table
    mode: insert_direct
    after_load: "UPDATE target_table ... SELECT ... FROM work_table ..."
</pre>

However, you need to truncate work_table every time before executing the yml.
If we add `truncate_insert_direct` mode, which will truncate the target table before `insert_direct` , it might be useful.

**Comment by Olaktal** [comment]
Hi @hito4t,

Oh yes you are right ! For the truncate, I'll drop the work table at the end of the SQL statement because I want my DB schema to be clean after import. 
Imagine I have like 40 tables to load so if each use a work table it will be a real mess ^^

**Comment by hito4t** [comment]
Hi @Olaktal ,
Now I've created the PR https://github.com/embulk/embulk-output-jdbc/pull/141 for the `after_load` feature.

**Comment by Olaktal** [comment]
Hi @hito4t,

Thanks a lot !
Could you please join to the PR the update of merge function ?

Thanks

**Comment by Olaktal** [comment]
I tried to selfupdate embulk but I got the following error:

```
ubuntu@ip-172-31-42-201:~$ embulk selfupdate
2016-11-04 15:02:29.955 +0000: Embulk v0.8.13
Checking the latest version...
Found new version 0.8.15.
Downloading https://dl.bintray.com/embulk/maven/embulk-0.8.15.jar ...
OpenSSL::SSL::SSLError: certificate verify failed
     connect_nonblock at org/jruby/ext/openssl/SSLSocket.java:225
              connect at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/net/http.rb:938
             do_start at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/net/http.rb:868
                start at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/net/http.rb:857
            open_http at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/open-uri.rb:319
          buffer_open at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/open-uri.rb:737
   block in open_loop at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/open-uri.rb:212
                catch at org/jruby/RubyKernel.java:1103
            open_loop at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/open-uri.rb:210
             open_uri at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/open-uri.rb:151
  block in selfupdate at /home/ubuntu/.embulk/bin/embulk!/embulk/command/embulk_selfupdate.rb:56
                 open at org/jruby/RubyIO.java:1131
           selfupdate at /home/ubuntu/.embulk/bin/embulk!/embulk/command/embulk_selfupdate.rb:54
                  run at /home/ubuntu/.embulk/bin/embulk!/embulk/command/embulk_run.rb:279
                <top> at /home/ubuntu/.embulk/bin/embulk!/embulk/command/embulk_main.rb:2
              require at org/jruby/RubyKernel.java:944
               (root) at uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:1
                <top> at file:/home/ubuntu/.embulk/bin/embulk!/embulk/command/embulk_bundle.rb:51
```

Do you have any idea where does the problem come from ?

Thanks

**Comment by hiroyuki-sato** [comment]
What OS are you using?

If you using Ubuntu, please try the following command

```
sudo curl -o /usr/local/share/ca-certificates/GTEGlRoot.crt https://www.cybertrust.ne.jp/sureserver/download/root_ca/GTEGlRoot.txt
sudo update-ca-certificates
```

https://github.com/embulk/embulk/issues/461#issuecomment-239729868

**Comment by Olaktal** [comment]
Hi @hiroyuki-sato,

Yes thanks it was ubuntu and it worked perfectly :)

**Comment by Olaktal** [comment]
@hiroyuki-sato 

As you may have noticed, I created a pull request for the update of the merge mode to support upsert.
Please keep me posted when it will be released.

Thanks

**Comment by hiroyuki-sato** [comment]
@Olaktal 
@hito4t may review it. 
However, your PR(#142) branch has conflicts that must be merged. 
You may need to resolve it.

**Comment by hito4t** [comment]
@Olaktal 
I'll resolve conflicts when merging.
No problem.

Now I'm testing.

**Comment by Olaktal** [comment]
Hi @hito4t @hiroyuki-sato 
What are the conflicts about ?
Thanks @hito4t for taking care of it :)

**Comment by hiroyuki-sato** [comment]
You and @hito4t modified `RedshiftOutputConnection.java` at same time. 

And the following part fails to auto-merge.

```
<<<<<<< 6e0f29a14daaf6985f658dab454f7701f63553bc
=======
import java.util.List;
import java.util.ArrayList;
>>>>>>> Update merge mode to work like Upsert
```

In this case, you may need to resolve the conflict.

What is `conflict`?

[Conflits de fusions](https://git-scm.com/book/fr/v2/Les-branches-avec-Git-Branches-et-fusions%C2%A0%3A-les-bases) (French)
[Basic-Merge-Conflicts](https://git-scm.com/book/en/v2/Git-Branching-Basic-Branching-and-Merging#Basic-Merge-Conflicts) (English)

Commonly, you may operate `rebase` .

Rebase
[Git Branching - Rebasing](https://git-scm.com/book/en/v2/Git-Branching-Rebasing)

If you want to know more detail. please let me know.
